### PR TITLE
Updates to docker-maven-plugin configuration so M1 mac build runs (almost) clean with `-Dstart-containers`

### LINKED ISSUE #25428: Improve podman compatibility on M1 for build test suites which use fabric8 docker-maven-plugin
### Describe the bug

https://github.com/quarkusio/quarkus/pull/25339 improved podman compatibility in some of the test suites which use fabric8 containers for databases. However, I'm still seeing failures on Mac M1 with podman. 

Once I worked out I needed to do `podman machine init -v $HOME:$HOME` for podman < 4.1, most of the issues I've been seeing relate to filesystem mounting and health checks. 

## Volume mounts 

The ':Z' SELinux access label seems to cause problems on M1. I think this is related to https://github.com/containers/podman/issues/13631. (`src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildLocalContainerRunner.java` also used :Z so I'm not sure why I didn't have to change anything there. )

## Health checks 

This is more troublesome. I'm struggling to get the docker maven plugin to successfully `<wait>` for mariadb to be up. I've tried a range of options. 

For example, I could not find a wait of doing a <wait> (from http://dmp.fabric8.io/#build-healthcheck) that didn't treat the `<time>` as both a minimum and a maximum. Several of the options I tried always passed, even if the container wasn't ready, or always failed, even if the container was actually ready. 

## What about test containers?

The fabric8 plugin is not under super-active development, and we're starting to see some non-ideal behaviours.  Longer term, we probably want to move away from the fabric8 docker-maven-plugin to test containers, but last time we looked into it, the fabric8 plugin gave us more flexibility to do things like define a sophisticated health check for container readiness, or inject a volume into the container with config to speed up startup. 

### Expected behavior

Tests should pass cleanly with `TESTCONTAINERS_RYUK_DISABLED="true" ./mvnw -Dquickly -DskipTests=false -Dstart-containers`

### SELinux access issues 

```
holly@hcummins-mac quarkus % MYSQL_USER=hibernate_orm_test MYSQL_PASSWORD=hibernate_orm_test MYSQL_DATABASE=hibernate_orm_test MYSQL_RANDOM_ROOT_PASSWORD=true podman run --env MYSQL\*  -v ./integration-tests/hibernate-orm-tenancy/connection-resolver/custom-mariadbconfig/:/etc/mysql/conf.d:Z docker.io/mariadb:10.7
Error: error preparing container 0c92a88aedfb270b2c2c81aa9e98b025170295b2f7e0a1b8ac42bd6492bd42b9 for attach: setxattr /Users/holly/Code/quarkus/myfork/quarkus/integration-tests/hibernate-orm-tenancy/connection-resolver/custom-mariadbconfig: operation not supported
```

### Health checks 

I'm seeing a range of problematic behaviours, which mostly involve waits reporting themselves as failed when they should be successful, or reporting as failed when they should be successful. 

### How to Reproduce?

```
TESTCONTAINERS_RYUK_DISABLED="true" ./mvnw -Dquickly -DskipTests=false -Dstart-containers -f extensions/reactive-mysql-client/deployment 
```
on M1 without docker installed (only podman)

### Output of `uname -a` or `ver`

Darwin hcummins-mac 21.4.0 Darwin Kernel Version 21.4.0: Fri Mar 18 00:46:32 PDT 2022; root:xnu-8020.101.4~15/RELEASE_ARM64_T6000 arm64

### Output of `java -version`

_No response_

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

_No response_

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by holly-cummins**
Fixes #25428. This is a big PR, sorry! I kept fixing-one-more-thing before doing the PR. I've left it unsquashed, partly because the history for each individual file is pretty clean and descriptive, and partly because I made the mistake of merging upstream into my long-running branch before squashing and so there's big merge commits in which are complicated to squash around. 

Here's what's changed: 

# DB2

- Set a platform override on M1 to pull the amd64 image (rather than trying to pull aarch64 and then dying). I’m not sure if this affects all M1 or just M1 + podman. (Most other database did not require the explicit platform flag, even if they did not have an arm64 platform-specific image, but some did. I don't fully understand why.)

We may want to do something similar in container-build-jib-with-db2 and also the dev services but I have not touched those.

# MariaDb

I had a problem with tcp-ping based readiness checks for mariadb. Only one test used it, so we can switch to using an `sqladmin ping` check. Several tests do a ping inside a `<postStart>` in the `<wait>`.  However, a `<postStart>` inside a `<wait>` isn’t a true readiness check. The plugin will always wait for the `<`time>` and then run the `<postStart>`. A better (but more verbose) option is to use a container health check. 

This was only strictly necessary where the tests were using a tcp ping to check for readiness (which did not work on podman), but I think it’s an improvement elsewhere. 

This sped up the tests quite a bit for me since it will cut the `<time>` wait short if the database is ready before the timeout pops.  I saw a factor of two improvement on tests which used mariadb, but on slower hardware the effect may not be so noticeable. 

Be aware when testing healthcheck scripts, caching of named images can cause apparently non-deterministic behaviour. Change the name if making changes. 

I also found that trying to extract common code to parent poms caused tests not to be able to connect to the containers. I don’t understand the reason - perhaps some crucial fabric8 state ends up in a target directory at the parent level?

# MSSQL

This one was challenging. The MS SQL image segfaults on ARM. The best option seemed to be the widely recommended workaround of using Azure Edge Server. The two products are not functionally interchangeable, but there is enough similarity to make it worth running this way. 

Note: this change will also affect dev services if built locally, but not from the official releases unless they are built on M1. I think that’s kind of weird but kind of ok, but I’m open to objections if people think that's too inconsistent.

# MYSQL 

Like DB2, I got a 404 if I tried to pull the image without explicitly specifying the architecture. However, just specifying the architecture wasn’t enough for the container to start. It would work on the command line with podman, but not in fabric8 with podman. 

Following https://www.emmanuelgautier.com/blog/mysql-docker-arm-m1 and https://stackoverflow.com/questions/65456814/docker-apple-silicon-m1-preview-mysql-no-matching-manifest-for-linux-arm64-v8 I instead changed the image name to one with better behaviour on arm. 

# Oracle

I hit https://github.com/fabric8io/docker-maven-plugin/issues/1369 with the log readiness check. Based on the discussion, I wonder if it’s a timing issue which my machine exposes by being fast. (I saw similar issues with `<log>` experimenting with the mariadb readiness check.)

ERRO[20873] accept tcp [::]:1521: use of closed network connection 
[ERROR] DOCKER> IO Error while requesting logs: org.apache.http.ConnectionClosedException: Premature end of chunk coded message body: closing chunk expected Thread-6

The image we use has a healthcheck configured, so we can just use that with fabric8. 

I also hit a more serious problem, which is that I could not get the database to start on M1. It seemed to be the same issue as https://stackoverflow.com/questions/68605011/oracle-12c-docker-setup-on-apple-m1. Others there reported success with qemu, which podman is using under the covers. I tried with podman 4.0.3 and podman 4.1, without success. 

Reluctantly, I disabled the tests. We should re-evaluate with later versions of the Oracle database in the future.

# Elasticsearch 

According to https://stackoverflow.com/questions/65962810/m1-mac-issue-bringing-up-elasticsearch-cannot-run-jdk-bin-java, we need a version bump to get M1 support, up to 7.10.2. I had to go up two minor increments to 7.11.0 on the logstash version to get arm64 working.

I notice we don’t have a centralised elasticsearch version, unlike other dev services. Is this something we want to change?

I also hit a comedy problem, which I'm noting here in case it helps others. I got a port conflict on port 5000. The process running on this port turns out to be an AirPlay server. You can deactivate it in System Preferences › Sharing, and unchecking AirPlay Receiver.

# Container Image Invoker 

I did not look at https://github.com/quarkusio/quarkus/issues/25230, so those tests still fail with podman on M1.

# Other issues 

I had to restart my podman machine to resolve some unexplained failures. I guess it had got tired. This could affect runs of the whole test suite if it runs end to end.

### CODE REVIEW

File: extensions/reactive-mssql-client/deployment/pom.xml [hunk 1]
```diff
@@ -114,6 +114,8 @@
                 </plugins>
             </build>
         </profile>
+
```
Comment (maxandersen) [inline extensions/reactive-mssql-client/deployment/pom.xml:117]: no diff ? remove from commit?

File: extensions/reactive-mysql-client/deployment/pom.xml [hunk 2]
```diff
@@ -168,29 +185,28 @@
                                             <date>default</date>
                                             <color>cyan</color>
                                         </log>
-                                        <!-- Speed things up a bit by not actually flushing writes to disk -->
+                                        Speed things up a bit by not actually flushing writes to disk -->
```
Comment (maxandersen) [inline extensions/reactive-mysql-client/deployment/pom.xml:188]: did you mean to remove this comment start?

Comment (holly-cummins) [reply extensions/reactive-mysql-client/deployment/pom.xml:188]: I did not. :)
I think my IDE is trolling me, since at one point I was running that test successfully. Will fix.

**Comment by maxandersen** [comment]
Changes looks sensible to me but @Sanne better off verifying -  and I don't have a M1 :/

I did notice lots of non meaningfull whitespace changes - not sure if our formatter being bad or just your IDE being configured differently :) either way nothing major but the line count/diff would be a lot less without them.

**Comment by holly-cummins** [comment]
I noticed that too about the whitespace and was annoyed by it. I thought I'd followed our IDE setup instructions accurately and I was getting a lot of line wrapping, so I made some more updates to my settings beyond what was in `CONTRIBUTING.MD`, but I probably should have gone back and manually reverted all of the wrapping that did happen. (I reverted some, but not all.) 

It might be worth spending an hour or two with a bunch of us in a room with a new starter, seeing if we've got compatible settings + docs across a few IDEs. I think XML is probably hardest to get right and consistent because the tooling support is most fragmented.

**Comment by Sanne** [comment]
looks great! I'll need to run it on my workstation at home though - sorry not possible today.

@yrodiere are you ok with the Elasticsearch version changes?

**Comment by yrodiere** [comment]
> @yrodiere are you ok with the Elasticsearch version changes?

7.11 is affected by at least one bug that is relevant to Hibernate Search: https://github.com/elastic/elasticsearch/pull/53127

While I don't think we'll experience that bug in Quarkus integration tests, I'd rather upgrade directly to 7.16, which is the version used in Hibernate Search 6.1 integration tests [see below, I'll send a PR to upgrade].

However, be aware that the version bumps in this PR are apparently in `integration-tests/logging-gelf/pom.xml` only. I have no opinion on those as I barely know about the use case. We'd have to ask @loicmathieu .

I'm surprised the build passes without bumping the version used for Hibernate Search integration tests, though?

> I notice we don’t have a centralised elasticsearch version, unlike other dev services. Is this something we want to change?

Please note that, at least when it comes to Elasticsearch you updated the configuration of (some) integration tests, not dev services.

`integration-tests/logging-gelf` apparently uses its own version, which I think we should change, indeed.

Other extensions rely on Maven properties defined in build-parent, and I think logging-gelf should do the same:

```
        <!-- Defaults for integration tests -->
        <elasticsearch-server.version>7.10.0</elasticsearch-server.version>
        <elasticsearch.image>docker.elastic.co/elasticsearch/elasticsearch-oss:${elasticsearch-server.version}</elasticsearch.image>
        <elasticsearch.protocol>http</elasticsearch.protocol>
        <opensearch-server.version>1.2.3</opensearch-server.version>
        <opensearch.image>docker.io/opensearchproject/opensearch:${opensearch-server.version}</opensearch.image>
        <opensearch.protocol>http</opensearch.protocol>
```

Elasticsearch dev services use a version hardcoded in their configuration class:

https://github.com/quarkusio/quarkus/blob/31c2513099eaa20bcf7f21ad54e093c47cb5b6ca/extensions/elasticsearch-rest-client-common/deployment/src/main/java/io/quarkus/elasticsearch/restclient/common/deployment/ElasticsearchDevServicesBuildTimeConfig.java#L32

That last part is indeed at odds with the more sophisticated solutions we have for relational datasources, which relies on a properties file using interpolation to inject Maven properties:

https://github.com/quarkusio/quarkus/blob/36f8e56ea9220833a8a35895fdd7ad02eb016a0f/extensions/devservices/postgresql/src/main/java/io/quarkus/devservices/postgresql/deployment/PostgresqlDevServicesProcessor.java#L72

https://github.com/quarkusio/quarkus/blob/b7fad49f8338fdeffff1fbe643dd2460b4e888a1/extensions/devservices/postgresql/src/main/resources/postgresql-devservice.properties#L1

... but it's consistent with Kafka dev services, for example:

https://github.com/quarkusio/quarkus/blob/10fb97a9db78a24226d719eb59eedbf0d596681e/extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaDevServicesBuildTimeConfig.java#L48

So to sum up, it's a mess :)

I'll send a PR to make the IT configuration more consistent, and upgrade to 7.16; feel free to merge this PR first and I'll handle the conflicts.

I also added a comment to #25486 so that we eventually solve the problem of Elasticsearch dev services.

> I also hit a comedy problem, which I'm noting here in case it helps others. I got a port conflict on port 5000

That's probably related to `logging-gelf` rather than Elasticsearch itself, as Elasticserach uses port 9200 (and sometimes 9300).

**Comment by geoand** [comment]
The rest client failures can be addressed by rebasing on `main`

**Comment by Sanne** [comment]
It's a bit big, and I'm struggling to follow the diffs because of the included merge commits. It's also conflicting now with `main`...

@holly-cummins you think you could rebase it, and perhaps split it in smaller PRs ? Many tests failed: it might be useful to narrow it down, and also merge things in smaller bites.

**Comment by yrodiere** [comment]
> @holly-cummins you think you could rebase it, and perhaps split it in smaller PRs

FWIW you can probably skip the part about Elasticsearch completely, now: #25663 was merged and it includes an upgrade to Elasticsearch 7.16, so more recent than what you did here and what you need for M1.

**Comment by holly-cummins** [comment]
OK, first smaller PR in for @Sanne, with the Maria DB changes: https://github.com/quarkusio/quarkus/pull/25805
It's still touching eight files, but it's the same change duplicated across all of the files. I'll wait until it's merged to look at MYSQL, MSSQL, Oracle, and Elasticsearch (with the latter hopefully being a no-op because of @yrodiere 's changes). Otherwise I'll get a big teetering pile of interconnected PRs, all trying to change the parent pom.

**Comment by Sanne** [comment]
@holly-cummins should we close this one?

**Comment by holly-cummins** [comment]
Not quite yet! One more sub-PR to go (DB2), which I'm testing locally now. And then I need to validate elasticsearch, and then do a final run through to make sure I didn't miss anything. This is what's gone in:

- https://github.com/quarkusio/quarkus/pull/25878
- https://github.com/quarkusio/quarkus/pull/25832
- https://github.com/quarkusio/quarkus/pull/25805
- https://github.com/quarkusio/quarkus/pull/25830

**Comment by holly-cummins** [comment]
Investigation continues. Now, the DB2 tests pass without any extra specifying the image name (perhaps I updated podman and that made a difference?). This is good news, because as @Sanne points out, there is no maven property called ${os.family}, and what comes out of ${os.name} is not useful for selecting docker images. 

However, the gelf tests are failing even with @yrodiere 's changes, so I need to figure out why.

**Comment by holly-cummins** [comment]
Noted re. DB2 moving out of the mainline test suite, @Sanne. I reckon even if we move them to a separate repo (or guard them some other way, like with a -Dslow profile), we'd want them to pass on M1. However, my fix was pretty ugly, since it had to explicitly set an architecture that the container provider should be setting for us. And since I now can't make the DB2 tests fail, I agree we should leave it. 

I have a feeling that maybe things are working locally now because of some caching (a bit like https://github.com/testcontainers/testcontainers-java/issues/5275 but for architectures?). If so, the DB2 M1 issues may rear their head again. But if things fail in an M1 Actions build, we can revisit this fix. 

And apart from the DB2-mystery-non-fix, and one more volume access modifier I had to add, and #25230, this PR has now been dissected and absorbed. So I'll close.

**Comment by Sanne** [comment]
> Noted re. DB2 moving out of the mainline test suite, @Sanne. I reckon even if we move them to a separate repo (or guard them some other way, like with a -Dslow profile), we'd want them to pass on M1.

Yes of course.
